timeoutreader_windows.go

```diff
@@ -0,0 +1,259 @@
+//go:build windows || test_alt_timeoutreader
+
+package terminal
+
+import (
+	"os"
+	"sync"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"fortio.org/log"
+	"fortio.org/safecast"
+	"golang.org/x/sys/windows"
+)
+
+const IsUnix = false
+
+type SystemTimeoutReader = TimeoutReaderWindows
+
+func NewSystemTimeoutReader(stream *os.File, timeout time.Duration) *TimeoutReaderWindows {
+	return NewTimeoutReaderWindows(stream, timeout)
+}
+
+type TimeoutReaderWindows struct {
+	handle              syscall.Handle
+	timeoutMilliseconds uint32
+	blocking            bool
+	ostream             *os.File
+	buf                 []byte
+	mut                 sync.Mutex
+	resizeChannel       chan WindowsResize
+}
+
+const fdwMode = windows.ENABLE_EXTENDED_FLAGS
+
+func NewTimeoutReaderWindows(stream *os.File, timeout time.Duration) *TimeoutReaderWindows {
+	if timeout < 0 {
+		panic("Timeout must be greater than or equal to 0")
+	}
+	// if we don't set console mode, the inputrecord struct's values become.... corrupted?
+	err := windows.SetConsoleMode(windows.Handle(stream.Fd()), uint32(fdwMode))
+	if err != nil {
+		// TODO: decide how to handle this
+	}
+	return &TimeoutReaderWindows{
+		handle:              syscall.Handle(os.Stdin.Fd()),
+		timeoutMilliseconds: safecast.MustConv[uint32](timeout.Milliseconds()),
+		blocking:            timeout == 0,
+		ostream:             stream,
+		resizeChannel:       make(chan WindowsResize, 1),
+	}
+}
+
+func (tr *TimeoutReaderWindows) IsClosed() bool {
+	return tr.ostream == nil
+}
+
+func (tr *TimeoutReaderWindows) Close() error {
+	var err error
+	if tr.blocking && tr.ostream != nil {
+		err = tr.ostream.Close()
+		tr.ostream = nil
+	}
+	return err
+}
+
+func (tr *TimeoutReaderWindows) RawMode() error {
+	return nil
+}
+
+func (tr *TimeoutReaderWindows) NormalMode() error {
+	return nil
+}
+
+func (tr *TimeoutReaderWindows) StartDirect() {
+}
+
+func (tr *TimeoutReaderWindows) ChangeTimeout(timeout time.Duration) {
+	tr.timeoutMilliseconds = safecast.MustConv[uint32](timeout.Milliseconds())
+}
+
+func (tr *TimeoutReaderWindows) ReadBlocking(p []byte) (int, error) {
+	// TODO: figure out why we need these two lines every time we ReadBlocking but not for ReadWithTimeout
+	fdwMode := windows.ENABLE_EXTENDED_FLAGS
+	_ = windows.SetConsoleMode(windows.Handle(tr.handle), uint32(fdwMode))
```

Comment:
that does seem expensive indeed and also not checking for error doesn't seem great